File: GolangSolutions/Array/M15_三数之和.go
/*
	@author Hyper
	@date 2023/07/09
	@file M15_三数之和.go
*/

package Array

import "sort"

func threeSum(nums []int) (res [][]int) {
	n := len(nums)
	sort.Ints(nums)
	for i, x := range nums {
		if x > 0 {
			break
		}
		if i > 0 && x == nums[i-1] {
			continue
		}
		l, r := i+1, n-1
		for l < r {
			sum := x + nums[l] + nums[r]
			if sum == 0 {
				res = append(res, []int{x, nums[l], nums[r]})
				for l < r && nums[l] == nums[l+1] {
					l++
				}
				for l < r && nums[r] == nums[r-1] {
					r--
				}
				l++
				r--
			} else if sum < 0 {
				l++
			} else {
				r--
			}
		}
	}

	return
}

File: LeetCode/src/Array/M15_三数之和.java
package Array;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author Hyper
 * @date 2023/07/09，2025/02/06，2025/03/21，2025/05/19
 * @file M15_三数之和.java
 * <p>
 * 思路
 * 排序+相向双指针
 * <p>
 * 注意题干中去重的要求
 */
public class M15_三数之和 {
    public List<List<Integer>> threeSum(int[] nums) {
        int n = nums.length;
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        for (int i = 0; i < n - 2; i++) {
            if (nums[i] > 0) {
                break;
            }
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            int j = i + 1, k = n - 1;
            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k];
                if (sum == 0) {
                    res.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    while (j < k && nums[j] == nums[j + 1]) {
                        j++;
                    }
                    while (j < k && nums[k] == nums[k - 1]) {
                        k--;
                    }
                    j++;
                    k--;
                } else if (sum > 0) {
                    k--;
                } else {
                    j++;
                }
            }
        }

        return res;
    }
}
